src/merkle/mmr/full.rs

```diff
@@ -0,0 +1,232 @@
+/// the position.
+///
+/// Note:
+/// The result is a tree position `p`, it has the following interpretations. $p+1$ is the depth of
+/// the tree, which corresponds to the size of a merkle proof for that tree. $2^p$ is equal to the
+/// number of leaves in this particular tree. and $2^(p+1)-1$ corresponds to size of the tree.
+pub(crate) const fn leaf_to_corresponding_tree(pos: usize, forest: usize) -> Option<u32> {
+    if pos >= forest {
+        None
+    } else {
+        // - each bit in the forest is a unique tree and the bit position its power-of-two size
+        // - each tree owns a consecutive range of positions equal to its size from left-to-right
+        // - this means the first tree owns from `0` up to the `2^k_0` first positions, where `k_0`
+        // is the highest true bit position, the second tree from `2^k_0 + 1` up to `2^k_1` where
+        // `k_1` is the second higest bit, so on.
+        // - this means the highest bits work as a category marker, and the position is owned by
+        // the first tree which doesn't share a high bit with the position
+        let before = forest & pos;
+        let after = forest ^ before;
+        let tree = after.ilog2();
+
+        Some(tree)
+    }
+}
+
+/// Return a bitmask for the bits including and above the given position.
+pub(crate) const fn high_bitmask(bit: u32) -> usize {
+    if bit > usize::BITS - 1 {
+        0
+    } else {
+        usize::MAX << bit
+    }
+}
+
+/// Return the total number of nodes of a given forest
+///
+/// Panics:
+///
+/// This will panic if the forest has size greater than `usize::MAX / 2`
+pub(crate) const fn nodes_in_forest(forest: usize) -> usize {
+    // - the size of a perfect binary tree is $2^{k+1}-1$ or $2*2^k-1$
+    // - the forest represents the sum of $2^k$ so a single multiplication is necessary
+    // - the number of `-1` is the same as the number of trees, which is the same as the number
+    // bits set
+    let tree_count = forest.count_ones() as usize;
+    forest * 2 - tree_count
+}
+
+// MMR
+// ===============================================================================================
+
+/// A fully materialized Merkle Mountain Range, with every tree in the forest and all their
+/// elements.
+///
+/// Since this is a full representation of the MMR, elements are never removed and the MMR will
+/// grow roughly `O(n**2)` in number of leaf elements.
+pub struct FullMMR {
+    /// Refer to the `forest` method documentation for details of the semantics of this value.
+    pub(super) forest: usize,
+
+    /// Contains every element of the forest.
+    ///
+    /// The trees are in postorder sequential representation. This representation allows for all
+    /// the elements of every tree in the forest to be stored in the same sequential buffer. It
+    /// also means new elements can be added to the forest, and merging of trees is very cheap with
+    /// no need to copy elements.
+    pub(super) nodes: Vec<Word>,
```

Comment:
I've imagined this stored as `Vec<MerkleTree>` - but `Vec<Word>` is probably much more efficient.

Reply:
If you mean a `Vec<MerkleTree>` as something similar to `Vec<Vec<Word>>`, with different allocations, that would super inefficient in comparison. The biggest difference here is the number of `Word` copies, hashing operations would be the same because the results can be copied over.

Just for curiosity here are the numbers for the `Vec<Vec<_>>`:

I'm going to compute the upper bound, assuming the merge of two trees first allocates a new empty tree, then copy the elements over, and computes the missing hash. Counting layers from root to leaves, elements of that layer would have been copied `layer` times. Which means the number of copies is:

```math
\sum_{l=0}^{k} (l+1)*2^{l}
```

which should be (no promises about this), this many operations:

```math
2^{k+1}*k + 1
```

While this implementation should be $2^{k+1}-1$ copies. This also has a nice property that it is pretty much an append-only data structure, should also be pretty good for writing into disk.

Reply:
I think the proposal was to leverage the [`MerkleTree`](https://github.com/0xPolygonMiden/crypto/blob/main/src/merkle/merkle_tree.rs#L11) data structure but I agree it will be less efficient.